src/python/geoclaw/topotools.py

```diff
@@ -1593,7 +1594,402 @@ def make_shoreline_xy(self, sea_level=0):
+    # The index of the topo object whose resolution is m
+    mfid = mtopoorder[m]
+    
+    # The parameters of the m resolution topo object
+    topoparam_mfid = [topo[mfid].extent[0], topo[mfid].extent[1], topo[mfid].extent[2],
+                      topo[mfid].extent[3], topo[mfid].delta[0], topo[mfid].delta[1]]
+    
+    # Innermost step of recursion reaches this point, only using coarsest topo grid
+    # compute directly
+    if m == mtopofiles - 1:
+        mark1 = intersection(domain, topoparam_mfid)
+        if mark1[0]:
+            integral = topointegral(mark1[2], topoparam_mfid, topo[mfid].Z, 1, 1)
+        else:
+            integral = 0.0
+    else:
+        int1 = recurintegral(domain, mtopoorder, mtopofiles, m+1, topo)
+
+        # Whether new topo object intersects the grid cell
+        mark2 = intersection(domain, topoparam_mfid)
+        
+        # The new topo object intersects the grid cell. Corrections are needed
+        if mark2[1] > 0:
+            int2 = recurintegral(mark2[2], mtopoorder, mtopofiles, m+1, topo)
+            int3 = topointegral(mark2[2], topoparam_mfid, topo[mfid].Z, 1, 1)
+            integral = int1 - int2 + int3
+        else:
+            integral = int1
+    return integral
+        
+
+def cellintegral(cell, mtopoorder, mtopofiles, topo):
+    r"""
+    Compute the integral on the surface defined by "cell" with the topo objects
+    list "topo".
+    
+    :Input:
+    - *cell* (ndarray(:)) The boundary of specific surface where the integral
+      is computed.
+    - *mtopoorder* (ndarray(:)) The order of the topo objects' resolutions.
+    - *mtopofiles* (int) The number of the topo objects.
+    - *topo* (list) The list of topo objects.
 
+    :Output:
+    - *topoint* (float) The integral on the cell, "cell".
+    """
+    
+    # Initialization
+    topoint = 0.0
+
+    for m in range(mtopofiles):
+    
+        # The index of the topo file object resolution is m
+        mfid = mtopoorder[m]
+        cellarea = (cell[1] - cell[0]) * (cell[3] - cell[2])
+        
+        # The parameters of this topo object
+        topoparam_mfid = [topo[mfid].extent[0], topo[mfid].extent[1], topo[mfid].extent[2],
+                          topo[mfid].extent[3], topo[mfid].delta[0], topo[mfid].delta[1]]
+        
+        # Whether m-th topo interects with cell
+        mark = intersection(cell, topoparam_mfid)
+        if mark[0]:
+            
+            # Whether the cell is completely overlapped by m-th topo
+            if mark[1] == cellarea:
+                topoint = topoint + topointegral(mark[2], topoparam_mfid, topo[mfid].Z, 1, 1)
+                return topoint
+            else:
+                
+                # If the cell is not completely overlapped by m-th topo, use recursion
+                # to compute the integral on the cell
+                topoint = recurintegral(cell, mtopoorder, mtopofiles, m, topo)
+                return topoint
+    return topoint
+
+
+class patch:
```

Comment:
Given the duplicate name in PyClaw and the fact that this is simply a data container can we change this to a dictionary?